### ISSUE TITLE
Sentry.init takes too much time and blocks ui thread

### ISSUE THREAD

**Issue by danikula** [getsentry/sentry-java#708]
On some android devices Sentry initialization may take up to 1-2 seconds 
what could be critical for a lot of apps.
<img width="236" alt="Sentry init" src="https://user-images.githubusercontent.com/1158827/57139117-031d2900-6dbd-11e9-87e3-ee25c2a3c04e.png">

The root of the problem is [static initialization block](https://github.com/getsentry/sentry-java/blob/b807ba22e1c651095d20a9f6d435586ff6591e21/sentry/src/main/java/io/sentry/config/Lookup.java#L35) in `Lookup` class which tries to find `sentry.properties` files in classpath. `Sentry` uses this `Lookup` class during initialization stage for getting initialization params from dsn. Even if all needed information is presented in dsn `Sentry` tries to load properties file because of `static` block that may take a lot of time.

@mcomella have already [mentioned](https://github.com/getsentry/sentry-java/issues/605#issuecomment-420765647) this problem in similar [issue](https://github.com/getsentry/sentry-java/issues/605).

**Comment by angusholder** [comment]
[This article](https://blog.nimbledroid.com/2016/04/06/slow-ClassLoader.getResourceAsStream.html#why-getresourceasstream-is-so-slow-in-android) explains why Android's resources are so slow.

I'm working around this problem by overriding all the methods of `AndroidSentryClientFactory` that use `Lookup.lookup` to instead just look in an instance of `Properties` which I store in my client factory class. 

A proper solution to this could be to add a `lookup` method to `SentryClientFactory` which `AndroidSentryClientFactory` could override to not try to find a `sentry.properties` file in resources, and instead access the properties through the assets mechanism. This is what the java.time backport for Android does to [avoid the slow resources mechanism when loading its timezone database](https://github.com/JakeWharton/ThreeTenABP/blob/master/threetenabp/src/main/java/com/jakewharton/threetenabp/AssetsZoneRulesInitializer.java#L24), so the properties would instead be accessed like `context.getAssets().open("sentry.properties")`

**Comment by justacodefan** [comment]
@angusholder since `Lookup` is only ever invoked by the factory or clients themselves, we might be able to not rely so much on the static context or accessors. I'll have something pushed up to address this soon.

**Comment by danikula** [comment]
@angusholder  Any news?

**Comment by ninniuz** [comment]
I am seeing a slowdown in the startup time of my apps as well due to this issue with Sentry.
@bruno-garcia do you think this can be addressed sooner or later?

**Comment by angusholder** [comment]
I worked around the problem by subclassing AndroidSentryClientFactory like [this](https://gist.github.com/angusholder/426b26f2890adfd349d4884f9ff9f1cf).

**Comment by bruno-garcia** [comment]
@ninniuz We haven't looked into this yet besides reading other's investigations.
Like the [issue linked at the top which was closed as not being an issue](https://github.com/mozilla-mobile/fenix/issues/2276#issuecomment-490281060).

Could I please have a repro on this? It seems to not be always reproducible otherwise this would be a blocker for many.

**Comment by ninniuz** [comment]
@bruno-garcia I can check on the worst devices I have to see if I can find one which takes that amount of time. 
On a Pixel 2 device though I see Sentry.init takes about 150 ms on average for a ~10MB apk.
I know it is not a lot of time, but this is essentially some wasted time.

**Comment by ninniuz** [comment]
> I worked around the problem by subclassing AndroidSentryClientFactory like [this](https://gist.github.com/angusholder/426b26f2890adfd349d4884f9ff9f1cf).

Thanks I will give this a try and compare the startup performance.

**Comment by ninniuz** [comment]
@bruno-garcia I was tracing the Sentry client init on a old Asus Z00AD device, Android Lollipop 5.0, with systrace, comparing the current performance and potential gains using the `SentryClientFactory` provided by @angusholder
I have tested on a proguarded and stripped down app, in release mode, which is ~10MB size.

What I see is a **~3x performance gain** on average, going from 1.5s down to about 500ms.

![Schermata 2019-07-11 alle 12 38 35](https://user-images.githubusercontent.com/3646303/61044849-d5b2b400-a3d9-11e9-8f98-fa2281525aa0.png)

![Schermata 2019-07-11 alle 12 35 25](https://user-images.githubusercontent.com/3646303/61045004-37731e00-a3da-11e9-9726-bfdf22a08f45.png)

Of course, there is some bits of performance degradation when tracing is enabled, but comparison should be accurate. I think this really might be worth taking a look.

**Comment by danikula** [comment]
As workaround for this problem I create an empty `sentry.properties` file and set path to it `sentry.properties.file` system property before Sentry initialization. It allows to prevent lookup `sentry.properties` in classpath.

**Comment by ninniuz** [comment]
@danikula can you please elaborate a bit more? I am not sure I got it.

**Comment by bruno-garcia** [comment]
You can set a ResourceLoader via `Sentry.init` now through options:

```
SentryOptions options = new SentryOptions();
options.ResourceLoader = new AndroidAssetsResourceLoader(ctx);
Sentry.init(options);
```

Feel free to reopen if there's any issue.

**Comment by angusholder** [comment]
Thanks for fixing this Bruno! Will be great to be able to remove my hacky workaround

**Comment by danikula** [comment]
It seems fix with manual setting resources loader doesn't work. Static initializer block in `Lookup` class runs  **before** resources loader was applied in `Sentry.init`.

**Comment by angusholder** [comment]
@bruno-garcia Looks like it hasn't quite worked, I think you want to switch the order of operations in `Sentry.init`, currently it looks like:

    SentryClient sentryClient = SentryClientFactory.sentryClient(
                sentryOptions.getDsn(),
                sentryOptions.getSentryClientFactory());
    Sentry.resourceLoader = sentryOptions.getResourceLoader();

But during initialization of the `SentryClient` instance it invokes `Lookup.lookup`, causing us to perform a lookup without the custom resource loader, so the performance bug remains

**Comment by bruno-garcia** [comment]
@angusholder yeah spot on. I'm fixing it on

**Comment by bruno-garcia** [comment]
Fixed pointed out by @angusholder was shipped version 1.7.27
https://search.maven.org/search?q=g:io.sentry
